piker/brokers/kucoin.py

```diff
@@ -0,0 +1,685 @@
+            ) as ws,
+            open_ping_task(ws, ping_interval, connect_id),
+        ):
+            log.info('Starting up quote stream')
+            # loop through symbols and sub to feedz
+            for sym in symbols:
+                pair: KucoinMktPair = pairs[sym]
+                kucoin_sym = pair.symbol
+
+                init_msgs = {
+                    # pass back token, and bool, signalling if we're the writer
+                    # and that history has been written
+                    sym: {
+                        'symbol_info': {
+                            'asset_type': 'crypto',
+                            'price_tick_size': float(pair.baseIncrement),
+                            'lot_tick_size': float(pair.baseMinSize),
+                        },
+                        'shm_write_opts': {'sum_tick_vml': False},
+                        'fqsn': sym,
+                    }
+                }
+
+
+                async with (
+                    subscribe(ws, connect_id, kucoin_sym),
+                    stream_messages(ws, sym) as msg_gen,
+                ):
+                    typ, quote = await anext(msg_gen)
+                    while typ != 'trade':
+                        # take care to not unblock here until we get a real trade quote
+                        typ, quote = await anext(msg_gen)
+
+                    task_status.started((init_msgs, quote))
+                    feed_is_live.set()
+
+                    async for typ, msg in msg_gen:
+                        await send_chan.send({sym: msg})
+
+@acm
+async def subscribe(ws: wsproto.WSConnection, connect_id, sym):
+    # level 2 sub
+    await ws.send_msg({
+        'id': connect_id,
+        'type': 'subscribe',
+        'topic': f'/spotMarket/level2Depth5:{sym}',
+        'privateChannel': False,
+        'response': True,
+    })
+
+    # watch trades
+    await ws.send_msg({
+        'id': connect_id,
+        'type': 'subscribe',
+        'topic': f'/market/ticker:{sym}',
+        'privateChannel': False,
+        'response': True,
+    })
+
+    yield
+
+    # unsub
+    if ws.connected():
+        log.info(f'Unsubscribing to {syn} feed')
+        await ws.send_msg(
+            {
+                'id': connect_id,
+                'type': 'unsubscribe',
+                'topic': f'/market/ticker:{sym}',
+                'privateChannel': False,
+                'response': True,
+            }
+        )
+
+
+@trio_async_generator
+async def stream_messages(ws: NoBsWs, sym: str) -> AsyncGenerator[NoBsWs, dict]:
+    timeouts = 0
+    last_trade_ts = 0
+
+    while True:
+        with trio.move_on_after(3) as cs:
+            msg = await ws.recv_msg()
+        if cs.cancelled_caught:
+            timeouts += 1
+            if timeouts > 2:
+                log.error('kucoin feed is sh**ing the bed... rebooting...')
+                await ws._connect()
+
+            continue
+
+        if msg.get('subject'):
```

Comment:
FYI (though not that you have to change it right now) you could actually match against your `msg` and then use object style matching for accessing all the values of the msg if you prefer:

https://jcristharif.com/msgspec/structs.html#pattern-matching

Reply:
ahh just to follow up, so that means you can use the object attributes as both a match condition, **and** an upacker system for binding to local variables in the scope:
https://peps.python.org/pep-0636/#adding-a-ui-matching-objects

Reply:
Although I do see the value of this, I'm not matching for anything more specific than the subject. I think we're fine to leave this for now.